runtime/tests/checker/external_mutation_test.go

```diff
@@ -0,0 +1,829 @@
+                    }
+                }
+            `, valueKind.Keyword(), access.Keyword(), declaration.Keywords(), assignmentOp, member.Code, destroyStatement),
+			)
+
+			if member.Mutating {
+				errs := ExpectCheckerErrors(t, err, 1)
+				var externalMutationError *sema.ExternalMutationError
+				require.ErrorAs(t, errs[0], &externalMutationError)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+
+	for _, access := range accessModifiers {
+		for _, kind := range declarationKinds {
+			for _, value := range valueKinds {
+				for _, member := range memberExpressions {
+					runTest(access, kind, value, member)
+				}
+			}
+		}
+	}
+}
+
+func TestCheckPubSetAccessModifier(t *testing.T) {
+
+	t.Parallel()
+	t.Run("pub set dict", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t,
+			`
+            pub contract C {
+                pub struct Foo {
+                    pub(set) var x: {Int: Int}
+            
+                    init() {
+                        self.x = {3: 3}
+                    }
+                }
+
+                pub fun bar() {
+                    let foo = Foo()
+                    foo.x[0] = 3
+                }
+            }
+        `,
+		)
+		require.NoError(t, err)
+
+	})
+}
+
+func TestCheckPubSetNestedAccessModifier(t *testing.T) {
+
+	t.Parallel()
+	t.Run("pub set nested", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t,
+			`
+            pub contract C {
+                pub struct Bar {
+                    pub let foo: Foo
+                    init() { 
+                       self.foo = Foo()
+                    }
+                }
+                
+                pub struct Foo {
+                    pub(set) var x: [Int]
+                
+                    init() {
+                       self.x = [3]
+                    }
+                }
+                
+                pub fun bar() {
+                    let bar = Bar()
+                    bar.foo.x[0] = 3
+                }
+            }
+        `,
+		)
+		require.NoError(t, err)
+
+	})
+}
+
+func TestCheckSelfContainingStruct(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("pub let", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t,
+			`
+            pub contract C {
+                pub struct Foo {
+                    pub let x: {Int: Int}
+            
+                    init() {
+                        self.x = {3: 3}
+                    }
+
+                    pub fun bar() {
+                        let foo = Foo()
+                        foo.x[0] = 3
+                    }
+                }
+            }
+        `,
+		)
+		require.NoError(t, err)
+
+	})
+}
+
+func TestCheckMutationThroughReference(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("pub let", func(t *testing.T) {
+
+		t.Parallel()
+
+		_, err := ParseAndCheck(t,
+			`
+            pub fun main() {
+                let foo = Foo()
+                foo.ref.arr.append("y")
```

Comment:
Shall we also add a test for taking a reference to the inner field, and updating it via reference?
i.e:
```
  var arrayRef = &foo.ref.arr as &[String]
  arrayRef[0] = "y"
```
(I guess the expected behavior is to not report errors?)